pex/pex_bootstrapper.py

```diff
@@ -56,28 +61,100 @@ def find_in_path(target_interpreter):
 
-  target_python = ENV.PEX_PYTHON
+
+def _handle_pex_python(target_python, compatibility_constraints):
+  """Re-exec using the PEX_PYTHON interpreter"""
   target = find_in_path(target_python)
   if not target:
     die('Failed to find interpreter specified by PEX_PYTHON: %s' % target)
+  elif compatibility_constraints:
+    pi = PythonInterpreter.from_binary(target)
+    if not all(pi.identity.matches(constraint) for constraint in compatibility_constraints):
+      die('Interpreter specified by PEX_PYTHON (%s) is not compatible with specified '
+          'interpreter constraints: %s' % (target, str(compatibility_constraints)))
   if os.path.exists(target) and os.path.realpath(target) != os.path.realpath(sys.executable):
     TRACER.log('Detected PEX_PYTHON, re-exec to %s' % target)
-    ENV.delete('PEX_PYTHON')
+    ENV.SHOULD_EXIT_BOOTSTRAP_REEXEC = True
     os.execve(target, [target_python] + sys.argv, ENV.copy())
 
 
+def _handle_general_interpreter_selection(pex_python_path, compatibility_constraints):
+  """Handle selection in the case that PEX_PYTHON_PATH is set or interpreter compatibility
+     constraints are specified.
+  """
+  compatible_interpreters = _find_compatible_interpreters(
+    pex_python_path, compatibility_constraints)
+
+  target = None
+  if compatible_interpreters:
+    target = min(compatible_interpreters).binary
+  if not target:
+    die('Failed to find compatible interpreter for constraints: %s'
+        % str(compatibility_constraints))
+  if os.path.exists(target) and os.path.realpath(target) != os.path.realpath(sys.executable):
+    if pex_python_path:
+      TRACER.log('Detected PEX_PYTHON_PATH, re-exec to %s' % target)
+    else:
+      TRACER.log('Re-exec to interpreter %s that matches constraints: %s' % (target,
+        str(compatibility_constraints)))
+    ENV.SHOULD_EXIT_BOOTSTRAP_REEXEC = True
+    os.execve(target, [target] + sys.argv, ENV.copy())
+
+
+def maybe_reexec_pex(compatibility_constraints=None):
+  if ENV.SHOULD_EXIT_BOOTSTRAP_REEXEC:
+    return
+  if ENV.PEX_PYTHON and ENV.PEX_PYTHON_PATH:
+    # both vars are defined, fall through to PEX_PYTHON_PATH resolution (give precedence to PPP)
+    pass
+  elif ENV.PEX_PYTHON:
+    # preserve PEX_PYTHON re-exec for backwards compatibility
+    _handle_pex_python(ENV.PEX_PYTHON, compatibility_constraints)
+
+  if not compatibility_constraints:
+    # if no compatibility constraints are specified, we want to match against
+    # the lowest-versioned interpreter in PEX_PYTHON_PATH if it is set
+    if not ENV.PEX_PYTHON_PATH:
+      # no PEX_PYTHON_PATH, PEX_PYTHON, or interpreter constraints, continue as normal
+      return
+
+  _handle_general_interpreter_selection(ENV.PEX_PYTHON_PATH, compatibility_constraints)
+
+
 def bootstrap_pex(entry_point):
   from .finders import register_finders
   register_finders()
-  maybe_reexec_pex()
+  pex_info = get_pex_info(entry_point)
+  with TRACER.timed('Bootstrapping pex with maybe_reexec_pex', V=2):
```

Comment:
the `get_pex_info` here should probably fall under the `TRACER.timed` block.

and a more concise/descriptive string here for the timer might be "Interpreter selection", probably with `V=3` to distinguish it in debug logging.